Regenerate password hash on a sign in after the stretches config changes

sergey-alekseev opened the pull request:
Closes #5143.  
Depends on #5092.

sergey-alekseev commented on lib/devise/strategies/database_authenticatable.rb:
```diff
@@ -13,6 +13,14 @@ def authenticate!
           resource.after_database_authentication
+
+          if BCrypt::Password.new(resource.encrypted_password).cost != resource.class.stretches
+            resource.update_column(
```
I did it intentionally.

sergey-alekseev commented on CHANGELOG.md:
```diff
@@ -1,6 +1,8 @@
 * enhancements
   * Increase default stretches to 12 (by @sergey-alekseev)
+  * Set stretches to the minimum cost supported by the bcrypt algorithm in tests (by @sergey-alekseev)
+  * Regenerate password hash on a sign in after the stretches config changes (by @sergey-alekseev)
```
It's consistent with the rest of the file.

sergey-alekseev commented on CHANGELOG.md:
```diff
@@ -1,6 +1,8 @@
 ### Unreleased
 * enhancements
   * Increase default stretches to 12 (by @sergey-alekseev)
+  * Set stretches to the minimum cost supported by the bcrypt algorithm in tests (by @sergey-alekseev)
```
It's consistent with the rest of the file.

sergey-alekseev commented on lib/devise/strategies/database_authenticatable.rb:
```diff
@@ -24,6 +25,17 @@ def authenticate!
+
+      def regenerate_bcrypt_password_hash(resource, password)
+        if !Devise.respond_to?(:encryptor) &&
+          ::BCrypt::Password.new(resource.encrypted_password).cost != resource.class.stretches
```
Will change during the next iteration when I get human feedback regarding the PR's code.

sergey-alekseev replied:
Done.

sergey-alekseev commented on lib/devise/strategies/database_authenticatable.rb:
```diff
@@ -24,6 +25,17 @@ def authenticate!
+      private
+
+      def regenerate_bcrypt_password_hash(resource, password)
+        if !Devise.respond_to?(:encryptor) &&
```
Looking forward for your recommendations how to better detect whether bcrypt or other hashing algorithm is used.

sergey-alekseev commented:
Some Travis CI jobs failed because of this bug – https://github.com/seattlerb/minitest/issues/803. The bug was introduced recently in the latest version. The fix is already in the minitest's master, just needs a gem version to be released.

I was able to reproduce it locally after installing Ruby 2.3 and doing `bundle update` (Travis [does this](https://github.com/plataformatec/devise/blob/f48b6f1651233a39c52dd70e9ac6c35f70086062/.travis.yml#L79) before running `bin/test`).

sergey-alekseev commented:
One Ruby 2.2 Travis CI test [failed](https://travis-ci.org/plataformatec/devise/jobs/590776409) because they released a gem version with both the match fix and Ruby 2.3+ requirement. So Ruby 2.2 is stuck with a buggy 5.12.0, not sure how to better fix it without additional Minitest gem updates. [Started](https://github.com/seattlerb/minitest/issues/803#issuecomment-536182589) a discussion in that issue.

sergey-alekseev commented:
✅ Travis CI is green now because a Minitest's maintainer released a new gem version.

mainameiz commented on lib/devise/strategies/database_authenticatable.rb:
```diff
@@ -24,6 +25,17 @@ def authenticate!
+            ::BCrypt::Password.new(resource.encrypted_password).cost != resource.class.stretches
+          resource.password = password
+          resource.skip_password_change_notification!
+          resource.save
```
Please pay attention here.
`update_column` should be used here, but sourcelevel-bot consider this as a mistake, but it isn't - https://github.com/plataformatec/devise/pull/5145#discussion_r328703342.

sergey-alekseev replied:
@mainameiz what do you mean by "Please pay attention here."?

Thanks for referencing my own comment regarding the usage of `update_column`. I used it initially and then during a refactoring iteration found the usage of `resource.skip_password_change_notification!` in the repo which (as far as I remember) in conjunction with `resource.save` seemed like a more proper implementation. Again, as far as I remember that's the only callback we need to avoid during the update. I haven't looked into the related repo code again now, but what I just wrote should be trustworthy :) I'd appreciate if you can review and verify.

mainameiz replied:
I'm not really familiar with this codebase but I suppose that `resource` variable is a instance of `User` model of some application. So it's state in database may be not valid. And if it's not valid then `resource.save` will not update the password.
I suppose, Devise should regenerate the digest nevertheless, so `update_column` should be used here.

sergey-alekseev replied:
Thanks for the feedback. Indeed, `resource.save` may fail if a `resource` is not valid.

I recalled why `#update_column` can't be used. See [the Travis CI build](https://travis-ci.org/plataformatec/devise/builds/590032290) for [the initial commit](https://github.com/plataformatec/devise/commit/496dfe9fe666707b190f4de42e1da19bc0d346ee) where `#update_column` was used. All jobs with `DEVISE_ORM=mongoid`  failed because it's an Active Record method.

msxavi replied:
What about `resource.save(validate: false)`?

sergey-alekseev replied:
From https://guides.rubyonrails.org/v6.0/active_record_callbacks.html#running-callbacks:  
>The following methods **trigger** callbacks:
...
save(validate: false)
...

On the contrary we want to avoid callbacks like sending the password change notification.

msxavi replied:
@sergey-alekseev 
> On the contrary we want to avoid callbacks like sending the password change notification

Invoking `skip_password_change_notification!` as the name tells, skips the callback you mentioned.

Your original code `resource.save` also triggers callbacks, so I didn't get your point.

msxavi replied:
Maybe this could solve it:
```suggestion
          update_column(:encrypted_password, BCrypt::Password.create(password, cost: resource.class.stretches).to_s)
```

sergey-alekseev replied:
Ah, I thought you want to use  
```ruby
resource.save(validate: false)
```  
instead of  
```ruby
resource.skip_password_change_notification!
resource.save
```
because of the previous context in this thread.  

Now I see your suggestion was to replace `resource.save` with `resource.save(validate: false)`. This totally makes sense. `resource.save(validate: false)` is extensively used in the codebase (e.g. look into https://github.com/plataformatec/devise/pull/4194). I made the change in https://github.com/plataformatec/devise/pull/5145/commits/31998a959d339b5352ad42861df2b2dbc1aef8f7#diff-d1ac35f7ea071d6c944aeb7ee4cd06cfR36 and covered this change in tests in https://github.com/plataformatec/devise/pull/5145/commits/31998a959d339b5352ad42861df2b2dbc1aef8f7#diff-0cd3b7071e3422d894ff8564d7f8edc9R128. Thanks for the suggestion! @msxavi

sergey-alekseev replied:
Not sure about your suggestion to use `#update_column` as [just above](https://github.com/plataformatec/devise/pull/5145#discussion_r352292285) I explained why it can't be used. @msxavi

msxavi replied:
Oh sure, no problem. I ended up using `update_column` because `save(validate: false)` didn't work quite well for me due to database constraints. Just to be clear, I'm trying to solve the stretches issue in the product I work on.

That's my final code which seems to work:
```
    return unless self.encrypted_password.present?
    if BCrypt::Password.new(self.encrypted_password).cost != self.class.stretches && persisted?
      update_column(:encrypted_password, BCrypt::Password.create(password, cost: self.class.stretches).to_s)
    end
```

sergey-alekseev replied:
Curious why `save(validate: false)` didn't work in your case.  

I use similar code with `#update_column` in a monkey-patched `Devise::Strategies::DatabaseAuthenticatable`, but this PR is to replace this monkey-patched solution.

msxavi commented:
How are we tracking with this?

sergey-alekseev commented:
> How are we tracking with this?

Is it a question to the devise owners? Can you please clarify your question? @msxavi

msxavi commented:
> > How are we tracking with this?
> 
> Is it a question to the devise owners? Can you please clarify your question? @msxavi

@sergey-alekseev since you are responsible for this PR, the question was meant for you.

sergey-alekseev commented:
>How are we tracking with this?

To be clear – I don't understand this question. @msxavi

aried3r commented:
Hey! We're interested in this feature as well, since we are currently using monkey-patches which we need to review again every time a new devise version is released and having this as part of the official code would be great!

Is there any way we can help this become a reality?

ghiculescu commented on lib/generators/templates/devise.rb:
```diff
@@ -103,15 +103,22 @@
+    # Limiting the stretches to the minimum cost supported by the algorithm
+    # will increase the performance of your test suite dramatically.
+    #
+    # For bcrypt, the minimum cost is 4. For other algorithms it defaults to 1.
```
> For bcrypt, the minimum cost is 4.

This is news to me and probably many others, given the old default of 1 here. Is this documented anywhere? What happens if you use a value less than 4?

sergey-alekseev replied:
Thanks for the questions. I thought the comment answers the question, but probably this should be clarified. Can you move your question to https://github.com/heartcombo/devise/pull/5092, so we proceed there? @ghiculescu

sergey-alekseev commented:
I rebased over the latest master to resolve minor merge conflicts and removed one commit as a dependency from this pull request. Originally I hoped for a quick merge, so I included a commit from #5092 to avoid merge conflicts.

sergey-alekseev commented:
Actually after excluding changes from https://github.com/heartcombo/devise/pull/5092 in this PR tests became inaccurate, so I had to apply [the changes above](https://github.com/heartcombo/devise/compare/6d18e1705406b8b0a73fe86bf0a941ec838d6749..80f65dda9ae4a5060c52b1248bb9b73a33de1734).

otherguy commented:
Would love this too!
